---
title: "Gene function analysis"
author: "Natalia Andrade and Ira Cooke"
date: "07/08/2017"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8,
                      echo=FALSE, warning=FALSE, message=FALSE)
knitr::opts_chunk$set(cache=TRUE)
options(width = 60)

library(knitr)
library(DESeq2)
library(tidyverse)
library(ggrepel)
library(pheatmap)
library(UpSetR)
library(ggplot2)
require(ggplot2); require(plyr); require(gridExtra); require(grid)
library(cowplot)

```



Load differentially expressed genes between HardCoral treatments obtain data from analysis with DESeq2 (03_deseq). 

```{r, include=FALSE}
# For the purpose of this analysis we first join 3 datasets together. 
# Dataset 1 is the differential expression statistics and vst transformed counts for all transcripts
# Dataset 2 is the annotation information for P.cylindrica transcripts
# Dataset 3 is a table of manual functional annotations for key genes

#
# Dataset 1
#
if (file.exists("cache/dds2.rds")) {
    dds2 <- read_rds("cache/dds2.rds")
} else {
  stop("Nothing in cache/dds2.rds. You must run 02_DESeq.Rmd first")
}

vsd2 <- varianceStabilizingTransformation(dds2, blind=FALSE)

contrast_ct <- list(c("HardCoralTrtPd_C","HardCoralTrtPf_C"),c("HardCoralTrtPd_T","HardCoralTrtPf_T"))

#Then we extract the results for the contrast between Control and Treatment
res_ct <- results(dds2,contrast = contrast_ct,listValues=c(1/2,
 -1/2), tidy = TRUE) %>% 
  arrange(padj) %>% 
  filter(padj<0.1) # Total 174 genes

#Transformed expression per sample (Dataset 1)
ct_res_vsd2 <- assay(vsd2)[which(rownames(vsd2) %in% res_ct$row),] %>%
  as.data.frame() %>%
  rownames_to_column("cluster_id") %>% 
  right_join(res_ct, by=c("cluster_id"="row"))
```

Annotating the 174 DEG_CvsT with transcriptome annotation and manual annotation
```{r annotating DEG_CvsT}
#
# Dataset 2
#
# Transcriptome annotation (location of the file temporary until in cloud)
P.cylindrica_144087anno<- readRDS("raw_data/P.cylindrica_144087anno.rda") %>% 
  ungroup() %>% 
  select(-PdLa,-PdLb, -PdLc,-PdLd,-PdLe, -PdC, -PfLa, -PfLb, -PfLc, -PfLd, -PfLe, -PfC)

# Dataset 3
#
# Manual annotation (also location temporary, it needs more edits don't put it in the cloud yet please)
Manual<- read.csv('raw_data/DE_193_database.csv')
```


```{r join_datasets}
DEG_174<- inner_join(P.cylindrica_144087anno,ct_res_vsd2, by=c("cluster_id")) %>% 
  left_join(Manual,by='cluster_id')
#write.csv(DEG_174, '/Users/jc275567/Dropbox/Mio/Competition/PORITES/Gene_Function_Analysis/DE_174_database.csv') #this data needs to be cleaned a bit more(Protein names)
```

Plots showing the expression of genes in the different categories and for the two genotypes

```{r}
sample_data <- read_csv("raw_data/Samples_data.csv",trim_ws = TRUE)
colnames(sample_data) <- c("ID","Tank","HardCoral",'treat',"SoftCoralControl","PdvsL_Other")

# Graph use in 
boxplot(log10(assays(dds2)[["cooks"]]), range=0, las=2)


# For genes DE ONLY in  contrast btw Control and treatment

top_genes <- DEG_174 %>% arrange(padj) %>% ungroup() %>% top_n(n=20, wt = padj) %>% pull(cluster_id)

DEG_174 %>%
  filter(cluster_id %in% top_genes) %>% select(cluster_id, PdLa,PdLb, PdLc,PdLe, PdC, PfLa, PfLb, PfLc, PfLd, PfLe, PfC) %>% 
  pivot_longer(c(-cluster_id),names_to = "ID",values_to="count") %>%
  left_join(sample_data) %>% 
  ggplot(aes(x=treat, y=log(count), colour=HardCoral)) +
  geom_point(aes(shape=SoftCoralControl)) +
  facet_wrap(~cluster_id,ncol = 5,scales = "free_y") +
  ggtitle("Top 20 differentially expressed genes in treatment (competition) versus control")

# For gene classify as RESISTING related genes

topResistingPorites<- DEG_174 %>% arrange(padj) %>% filter(RESISTING==TRUE) %>% select(cluster_id, PdLa,PdLb, PdLc,PdLe, PdC, PfLa, PfLb, PfLc, PfLd, PfLe, PfC) 

plot_portiesR<- topResistingPorites %>%  
    pivot_longer(c(-cluster_id),names_to = "ID",values_to="count") %>%
    left_join(sample_data) %>% 
    ggplot(aes(x=treat, y=log(count), colour= HardCoral)) + 
    geom_boxplot() + 
    facet_wrap( ~ cluster_id ,scales = "free_y")+
  ggtitle("Differentially expressed genes in functional category (RESISTANCE) in treatment versus control")

plot_portiesR

# For genes classify as CELLULAR STRESS related

topCellularstressPorites<- DEG_174 %>% arrange(padj) %>% filter(CELLULAR.STRESS==TRUE) %>% select(cluster_id, PdLa,PdLb, PdLc,PdLe, PdC, PfLa, PfLb, PfLc, PfLd, PfLe, PfC) 
plot_portiesCS<- topCellularstressPorites %>%  
    pivot_longer(c(-cluster_id),names_to = "ID",values_to="count") %>%
    left_join(sample_data) %>% 
    ggplot(aes(x=treat, y=log(count), colour= HardCoral)) + 
    geom_boxplot() + 
    facet_wrap( ~ cluster_id ,scales = "free_y") +
  ggtitle("Differentially expressed genes in functional category (CELLULAR STRESS) in treatment versus control")
plot_portiesCS
```

Ploting expression of Lobophytum DEG and Porites DEG to show effect of colony Pd
```{r}
# First prepair Lobophytum data
Trans_expr<- read.delim('raw_data/Lc_LvsPd_all_0.5_exp.txt', header = TRUE, sep= "\t", row.names = NULL)

Trans_expr<- dplyr::rename(Trans_expr,LdPe = L10_S10.bam, LdPf= L11_S11.bam, LdPd= L12_S12.bam, LePe= L13_S13.bam, Le= L14_S14.bam, LePf=L15_S15.bam, LePd= L16_S16.bam, LcPe= L17_S10.bam, Lc= L18_S11.bam, LcPf= L19_S12.bam, LaPd= L1_S1.bam, LcPd= L20_S13.bam, La= L2_S2.bam, LaPf=L3_S3.bam, LaPe= L4_S4.bam, Lb= L5_S5.bam, LbPe=  L6_S6.bam, LbPd= L7_S7.bam, LbPf= L8_S8.bam, Ld= L9_S9.bam)
names(Trans_expr)
Trans_expr<- dplyr::select(Trans_expr,Cluster,LaPd, LcPd,LdPd, LePd,LbPd,LaPe, LbPe, LcPe, LdPe, LePe, LaPf, LbPf, LcPf, LdPf, LePf, La, Lb, Lc, Ld, Le)

Trans_expr<- dplyr::tbl_df(Trans_expr)

#DEG in Lobophytum under competition compared to Lobophytum control
DE_339<- read.csv('raw_data/DEG_339_LobophytumComp_20190116.csv')

#Lobophytum expression don't select Pe, Top50 genes
Exp_data_Lobo<- left_join(DE_339, Trans_expr, by=c("name"="Cluster")) %>% select(name,LaPd,LcPd,LdPd,LePd,LbPd,LaPf,LbPf,LcPf,LdPf,LePf,La,Lb,Lc,Ld,Le)

top50Lobo<- left_join(DE_339, Trans_expr, by=c('name'='Cluster')) %>% arrange(padj) %>% top_n(50) %>% select(name,LaPd,LcPd,LdPd,LePd,LbPd,LaPf,LbPf,LcPf,LdPf,LePf,La,Lb,Lc,Ld,Le) 

#Top 50 genes DE for Porites
top50Porites<- DEG_174 %>% arrange(padj) %>% top_n(50) %>% select(cluster_id, PdLa,PdLb, PdLc,PdLe, PdC, PfLa, PfLb, PfLc, PfLd, PfLe, PfC) 

#Porites expression and Lobophytum
sample_data_both<- read.csv('raw_data/Samples_data_both.csv')
Exp_PoritesLobo<- full_join(top50Porites, top50Lobo, by=c('cluster_id'='name'))

plot_ExpPoritesLobo<- Exp_PoritesLobo %>% as.data.frame() %>%
pivot_longer(c(-cluster_id),names_to = "ID",values_to="count")  %>% 
  filter(!is.na(count)) %>%
  left_join(sample_data_both) %>% 
  ggplot(aes(x=treat , y=log(count), colour=HardCoralControl )) +
    geom_boxplot() +
    facet_grid(Coral~SoftCoralControl,scales = "free_y")
plot_ExpPoritesLobo
```

Manual Categories plots
```{r, gene function }
DE_174_database<- read.csv('~/Dropbox/natalia_competition/Porites_competition/gene_function/data/DE_174_database_underEdition2020.csv')
#(location of the document needs to change)

# Upset plots can be made much nicer (show u and than reg proportions, show which ones have this strong difference in expression)

upset(DE_174_database, sets= c("Sensory.Nervous","Immunity","Cellular.Stress"), 
         order.by= c('freq','degree'),
         keep.order= TRUE,
         sets.bar.color("orange"), 
         point.size=2,
         main.bar.color = "maroon",
         matrix.color = "gray23",
         text.scale =2)  
 
#CELLULAR.STRESS #IMMUNE RESPONSE TO COMPETITION
upset(DE_174_database, sets= c("Toxins.defense","Apoptosis",'Immune.activation',
                               "Ubiquitination","Mucus","Antioxidants","Immunity","Cellular.Stress"), 
         order.by= c('freq','degree'), 
         keep.order= TRUE,
         sets.bar.color("orange"), 
         point.size=2,
         main.bar.color = "maroon",
         matrix.color = "gray23",
         text.scale =2) 

#NERVOUS SYSTEM AND BEHAVIOUR
upset(DE_174_database, sets= c("Behaviour","Muscle","GPCRs","Food.related","Autophagy","Sensory.Nervous"), 
      order.by= c('freq','degree'), 
      keep.order= TRUE,
      sets.bar.color("orange"), 
      point.size=2,
      main.bar.color = "maroon",
      matrix.color = "gray23",
      text.scale =2) 

```


